File: sorts/CountingSort.java
/**
 * Time Best: n+k Average: n+k Worst: n+k
 *
 * Space Worst: k
 */
public class Solution {

    public static int[] countingSort(int array[], int maxNumberValue) {
        int numberCounts[] = new int[maxNumberValue];
        int sortedArray[] = new int[array.length];
        int sortedIndex = 0;

        // Count how many instances of a number and place that
        // number in its corresponding index.
        for (int number : array) {
            numberCounts[number]++;
        }

        // Loop through each number count index.
        // Take the count of that number.
        for (int number = 0; number < numberCounts.length; number++) {
            int count = numberCounts[number];

            for (int i = 0; i < count; i++) {
                sortedArray[sortedIndex] = number;
                sortedIndex++;
            }
        }

        return sortedArray;
    }

    public static int findMaxValue(int data[]) {
        int max = 0;
        for (int value : data) {
            if (max < value) {
                max = value;
            }
        }
        return max;
    }

    public static int[] generateArray(int size) {
        int array[] = new int[size];
        for (int i = 0; i < array.length; i++) {
            array[i] = (int) (Math.random() * 90) + 10;
        }
        return array;
    }

    public static void main(String[] args) {
        int array[] = generateArray(20);
        int result[] = countingSort(array, findMaxValue(array) + 1);
        for (int number : result) {
            System.out.print(number + " ");
        }
    }
}

File: sorts/javascript/counting-sort.js

const countingSort = (data, maxValue) => {
    let countArray = [];
    let resultArray = [];
    let sortIndex = 0;

    for (const number of data) {
        if (countArray[number] === undefined) {
            countArray[number] = 0;
        }
        countArray[number]++;
    }

    for (let number = 0; number < countArray.length; number++) {
        let count = countArray[number];
        for (let i = 0; i < count; i++) {
            resultArray[sortIndex] = number;
            sortIndex++;
        }
    }

    return resultArray;
}

const getMaxNumber = (data) => {
    let max = 0;
    for (const value of data) {
        if (max < value) {
            max = value;
        }
    }
    return max;
}

let data = [3, 4, 1, 2, 12, 11, 9, 10];
let max = getMaxNumber(data);
let result = countingSort(data, max);
console.log([...result]);
